lib/liquid/standardfilters.rb

```diff
@@ -89,13 +89,22 @@ def truncate(input, length = 50, truncate_string = "...")
 
     def truncatewords(input, words = 15, truncate_string = "...")
       return if input.nil?
-      wordlist = input.to_s.split
-      words    = Utils.to_integer(words)
+      words = Utils.to_integer(words)
 
-      l = words - 1
-      l = 0 if l < 0
+      l = words
+      l = 1 if l <= 0
+
+      # Scan for non-space characters followed by one or more space characters
+      # `l` times.  Also ignore leading whitespace
+      str = input[/\A[ ]*(?:[^ ]*[ ]+){#{l}}/]
```

Comment:
The previous code was using `input.to_s.split`, which would split on more than just the space character, but now you are only splitting on the space character.  That seems like a breaking change.

Why are you using `*` to match on the non-whitespace characters?  Also, you are using `+` to match on the whitespace after the word, which makes the number of words that it will match dependant on trailing whitespace.  Also, why are you using an exact number of repetition match (`{#{l}}`), instead of an at most number of repetition match (`{,#{l}}`)?

Reply:
`String#split` docs say

> If pattern is nil, the value of $; is used. If $; is nil (which is the default), str is split on whitespace as if ' ' were specified.

... which is a bit ambiguous.

The exact repetition match works because in cases where there are fewer repetitions we don't want to apply truncation.

Reply:
> The exact repetition match works because in cases where there are fewer repetitions we don't want to apply truncation. There is a bit of odd behaviour with the trailing whitespace though (`"foo bar " | truncatewords: 2 => "foo bar..."`) good catch.

Oh, that makes sense, we want to avoid a match for the fast path of not truncating to avoid the overhead of allocating the match.  We might want a positive lookahead to make sure there is another word that will be truncated.